[SPARK-12799] Simplify various string output for expressions

liancheng opened the pull request:
This PR introduces several major changes:
1. Replacing `Expression.prettyString` with `Expression.sql`
   
   The `prettyString` method is mostly an internal, developer faced facility for debugging purposes, and shouldn't be exposed to users.
2. Using SQL-like representation as column names for selected fields that are not named expression (back-ticks and double quotes should be removed)
   
   Before, we were using `prettyString` as column names when possible, and sometimes the result column names can be weird.  Here are several examples:
   
   | Expression | `prettyString` | `sql` | Note |
   | --- | --- | --- | --- |
   | `a && b` | `a && b` | `a AND b` |  |
   | `a.getField("f")` | `a[f]` | `a.f` | `a` is a struct |
3. Adding trait `NonSQLExpression` extending from `Expression` for expressions that don't have a SQL representation (e.g. Scala UDF/UDAF and Java/Scala object expressions used for encoders)
   
   `NonSQLExpression.sql` may return an arbitrary user facing string representation of the expression.

liancheng committed:
Replaces Expression.prettyString with Expression.sql

liancheng committed:
Fixes Catalyst test failures

liancheng committed:
SQLBuilder should reject logical plans containing `NonSQLExpression`s

liancheng commented:
(I'd expect multiple test failures due to changes in exception error messages.)

liancheng commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:
```diff
@@ -233,7 +232,7 @@ private[hive] case class HiveGenericUDF(
 
   @transient
-  private lazy val deferedObjects = argumentInspectors.zip(children).map { case (inspect, child) =>
+  private lazy val deferredObjects = argumentInspectors.zip(children).map { case (inspect, child) =>
```
The original field name is misspelled here.

liancheng commented:
cc @rxin

SparkQA commented:
**[Test build #49408 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49408/consoleFull)** for PR 10757 at commit [`6a9f4bd`](https://github.com/apache/spark/commit/6a9f4bd701d9b966760cf414b319249f02dcbe2f).

SparkQA commented:
**[Test build #49408 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49408/consoleFull)** for PR 10757 at commit [`6a9f4bd`](https://github.com/apache/spark/commit/6a9f4bd701d9b966760cf414b319249f02dcbe2f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49408/
Test FAILed.

liancheng committed:
Fixes test failures

SparkQA commented:
**[Test build #49415 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49415/consoleFull)** for PR 10757 at commit [`226b884`](https://github.com/apache/spark/commit/226b884f86b844a6fd3c6f54bbd2aafb225909aa).

rxin commented:
High level changes LGTM

SparkQA commented:
**[Test build #49415 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49415/consoleFull)** for PR 10757 at commit [`226b884`](https://github.com/apache/spark/commit/226b884f86b844a6fd3c6f54bbd2aafb225909aa).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49415/
Test FAILed.

liancheng committed:
Fixes more test failures

SparkQA commented:
**[Test build #49429 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49429/consoleFull)** for PR 10757 at commit [`36ba05f`](https://github.com/apache/spark/commit/36ba05f2eaf2bf6752cc49246b0e11563dac3661).

SparkQA commented:
**[Test build #49429 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49429/consoleFull)** for PR 10757 at commit [`36ba05f`](https://github.com/apache/spark/commit/36ba05f2eaf2bf6752cc49246b0e11563dac3661).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49429/
Test FAILed.

rxin commented:
almost there...

one question: is ` a valid character for column name?

rxin commented:
(I'm thinking we should remove all backticks when generating name aliases)

liancheng commented:
I should add a `quote` method somewhere to handle backtick escaping (Hive escapes backticks within identifiers to double-backticks). By "name aliases", do you mean those like generated subquery names?

liancheng committed:
Fixes PySpark test failures

SparkQA commented:
**[Test build #49468 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49468/consoleFull)** for PR 10757 at commit [`f263cf5`](https://github.com/apache/spark/commit/f263cf569517c5110962f21c03adaebe50665952).

rxin commented:
I was referring to the column names automatically generated by spark sql. It is just taken from the "sql" representation of the column names right now, which includes backticks and can be a problem.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49473/
Test FAILed.

SparkQA commented:
**[Test build #49468 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49468/consoleFull)** for PR 10757 at commit [`f263cf5`](https://github.com/apache/spark/commit/f263cf569517c5110962f21c03adaebe50665952).
 * This patch **fails PySpark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49468/
Test FAILed.

yhuai commented:
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/interfaces.scala#L171 has the extra space.

liancheng committed:
Only quotes SQL identifiers when necessary

SparkQA commented:
**[Test build #49538 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49538/consoleFull)** for PR 10757 at commit [`9999f26`](https://github.com/apache/spark/commit/9999f2682bb55dedf29cfa7390dba03df72d24f9).

SparkQA commented:
**[Test build #49538 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49538/consoleFull)** for PR 10757 at commit [`9999f26`](https://github.com/apache/spark/commit/9999f2682bb55dedf29cfa7390dba03df72d24f9).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait ExtractValue `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49538/
Test FAILed.

liancheng commented:
A summary of offline discussion with @rxin and @marmbrus:

The reason why @rxin suggested removing all back-ticks in generated column names is mostly because of backwards compatibility. Those generated column names are now generated using `Expression.sql` instead of `Expression.prettyString`. For example, the following DataFrame

```scala
df.selectExpr("id + 1")
```

used to produce a single column named `id + 1`, but now it becomes `` `id` + 1``.

However, later on I found that removing back-ticks in generated column names still cannot guarantee this level of backwards-compatibility. This is because, although `pettyString` and `sql` are often quite similar, they are still inherently different from each other in many cases, for example:

Expression | `prettyString` | `sql`
---------- | -------------- | -----
`a && b` | `a && b` | `a AND b`
`a.getField("f")` | `a[f]` | `a.f`
`m.getItem("key")` | `m[key]` | `m["key"]`

Basically we won't be able to replace `prettyString` with `sql` is we do want this level of backwards-compatibility.  However, we may see that the original `prettyString` method doesn't return proper column names.  I tend to fix them in Spark 2.0.  As for back-tick quoting, a utility method `safeSQLIdent` is added to quote identifiers only when necessary.

SparkQA commented:
**[Test build #49544 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49544/consoleFull)** for PR 10757 at commit [`9fa6307`](https://github.com/apache/spark/commit/9fa6307ec2ec8c9432b1559ba76a2987d15d9b94).

SparkQA commented:
**[Test build #49544 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49544/consoleFull)** for PR 10757 at commit [`9fa6307`](https://github.com/apache/spark/commit/9fa6307ec2ec8c9432b1559ba76a2987d15d9b94).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait ExtractValue `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49544/
Test FAILed.

liancheng committed:
Handles SQL keywords when quoting SQL identifiers

SparkQA commented:
**[Test build #49562 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49562/consoleFull)** for PR 10757 at commit [`602a8eb`](https://github.com/apache/spark/commit/602a8ebc9bf0582305c2a907ae447ef6925e50dd).

SparkQA commented:
**[Test build #49562 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49562/consoleFull)** for PR 10757 at commit [`602a8eb`](https://github.com/apache/spark/commit/602a8ebc9bf0582305c2a907ae447ef6925e50dd).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49562/
Test FAILed.

SparkQA commented:
**[Test build #49593 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49593/consoleFull)** for PR 10757 at commit [`b7fb4b1`](https://github.com/apache/spark/commit/b7fb4b1577bfc2f53fcf8187d5974e4044f1ea97).

SparkQA commented:
**[Test build #49593 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49593/consoleFull)** for PR 10757 at commit [`b7fb4b1`](https://github.com/apache/spark/commit/b7fb4b1577bfc2f53fcf8187d5974e4044f1ea97).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49593/
Test FAILed.

SparkQA commented:
**[Test build #49594 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49594/consoleFull)** for PR 10757 at commit [`4cb6410`](https://github.com/apache/spark/commit/4cb64104b961e7a3c06e973d26810b0076f1ac3c).

SparkQA commented:
**[Test build #49594 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49594/consoleFull)** for PR 10757 at commit [`4cb6410`](https://github.com/apache/spark/commit/4cb64104b961e7a3c06e973d26810b0076f1ac3c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49594/
Test FAILed.

SparkQA commented:
**[Test build #49599 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49599/consoleFull)** for PR 10757 at commit [`9047dc5`](https://github.com/apache/spark/commit/9047dc51c049e5b6c10ecfdb972bc59db574c500).

SparkQA commented:
**[Test build #49599 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49599/consoleFull)** for PR 10757 at commit [`9047dc5`](https://github.com/apache/spark/commit/9047dc51c049e5b6c10ecfdb972bc59db574c500).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49599/
Test FAILed.

rxin commented:
@liancheng why do we need the quote parameter for the sql function? Can't we just replace the attribute references with PrettyAttribute, which does not include backticks?

marmbrus commented:
+1 to doing this with a transformation instead of threading a boolean all the way through.

liancheng commented:
Ah, that's a good point! Totally forgot `PrettyAttribute`.

liancheng committed:
Uses PrettyAttribute when generating column names

SparkQA commented:
**[Test build #49960 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49960/consoleFull)** for PR 10757 at commit [`eb78f29`](https://github.com/apache/spark/commit/eb78f29eea1cc84a9dafc63f65cf8bfc4aaa3243).

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49959/
Test FAILed.

SparkQA commented:
**[Test build #49960 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49960/consoleFull)** for PR 10757 at commit [`eb78f29`](https://github.com/apache/spark/commit/eb78f29eea1cc84a9dafc63f65cf8bfc4aaa3243).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49960/
Test FAILed.

liancheng committed:
Fixes test failures

liancheng committed:
Fixes some test failures

SparkQA commented:
**[Test build #50130 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50130/consoleFull)** for PR 10757 at commit [`df29c4f`](https://github.com/apache/spark/commit/df29c4fa1738de869c330fa4c4578a01c3742b4a).

SparkQA commented:
**[Test build #50130 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50130/consoleFull)** for PR 10757 at commit [`df29c4f`](https://github.com/apache/spark/commit/df29c4fa1738de869c330fa4c4578a01c3742b4a).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50130/
Test FAILed.

liancheng committed:
Fixes Scala style check

SparkQA commented:
**[Test build #50132 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50132/consoleFull)** for PR 10757 at commit [`726dba9`](https://github.com/apache/spark/commit/726dba90e248d5ef95cc8dfa03b259b07edb6083).

SparkQA commented:
**[Test build #50132 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50132/consoleFull)** for PR 10757 at commit [`726dba9`](https://github.com/apache/spark/commit/726dba90e248d5ef95cc8dfa03b259b07edb6083).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50132/
Test FAILed.

liancheng committed:
Fixes more test failures

SparkQA commented:
**[Test build #50143 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50143/consoleFull)** for PR 10757 at commit [`ff6424f`](https://github.com/apache/spark/commit/ff6424fd3f886c58c02ec7b491878ff4f6b86b2a).

liancheng committed:
Fixes PySpark test failures

liancheng committed:
Fixes math expression SQL representation

SparkQA commented:
**[Test build #50159 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50159/consoleFull)** for PR 10757 at commit [`9c7e545`](https://github.com/apache/spark/commit/9c7e5457219007252f6a8d6ba2271f4ad98a48db).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50159/
Test FAILed.

SparkQA commented:
**[Test build #50143 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50143/consoleFull)** for PR 10757 at commit [`ff6424f`](https://github.com/apache/spark/commit/ff6424fd3f886c58c02ec7b491878ff4f6b86b2a).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50143/
Test FAILed.

rxin commented:
I'd actually not have "PythonUDF#" as part of the name.

liancheng committed:
Fixes PythonUDF and R test failures

liancheng commented:
It took me sometime to fix those R test failures. Hopefully the last commit fixes everything (finger crossed).

liancheng commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala:
```diff
@@ -131,17 +131,17 @@ abstract class OrcSuite extends QueryTest with TestHiveSingleton with BeforeAndA
+        |  CAST(null as TIMESTAMP) as c7,
+        |  CAST(null as DATE) as c8,
+        |  CAST(null as STRING) as c9,
+        |  CAST(null as VARCHAR(10)) as c10
```
Aliases are added mostly for the last two fields: `VARCHAR` is equivalent to `STRING` for Spark SQL, so the last two fields actually have exactly the same column names.

SparkQA commented:
**[Test build #50308 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50308/consoleFull)** for PR 10757 at commit [`cc7dace`](https://github.com/apache/spark/commit/cc7dacecdfe2fc0966759992ed0610f66e04dbb3).

SparkQA commented:
**[Test build #50308 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50308/consoleFull)** for PR 10757 at commit [`cc7dace`](https://github.com/apache/spark/commit/cc7dacecdfe2fc0966759992ed0610f66e04dbb3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50308/
Test FAILed.

SparkQA commented:
**[Test build #2471 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2471/consoleFull)** for PR 10757 at commit [`cc7dace`](https://github.com/apache/spark/commit/cc7dacecdfe2fc0966759992ed0610f66e04dbb3).

SparkQA commented:
**[Test build #2471 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2471/consoleFull)** for PR 10757 at commit [`cc7dace`](https://github.com/apache/spark/commit/cc7dacecdfe2fc0966759992ed0610f66e04dbb3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala:
```diff
@@ -39,11 +39,11 @@ case class ScalaUDF(
   override def nullable: Boolean = true
 
-  override def toString: String = s"UDF(${children.mkString(",")})"
+  override def toString: String = s"ScalaUDF(${children.mkString(",")})"
```
Is this change necessary? It is really strange to call this ScalaUDF because Java also goes through this.

rxin replied:
We should rename the case class too - but that should be done separately.

liancheng replied:
Oh right, I didn't realize that Java UDF also goes here. Will revert this one.

liancheng committed:
Reverts ScalaUDF change

SparkQA commented:
**[Test build #50353 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50353/consoleFull)** for PR 10757 at commit [`eef9423`](https://github.com/apache/spark/commit/eef94237ade63e03ae698870b34b620bf1c88611).

rxin commented:
A few comments:

1. As discussed offline, let's be consistent and always quote in sql query generation, and never allow backticks in aliases.

2. Rename safeSQLIdent -> quoteIdentifier

3. Please add a unit test case for each expression that requires safeSQLIdent

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:
```diff
@@ -298,7 +299,14 @@ case class AttributeReference(
 case class PrettyAttribute(name: String, dataType: DataType = NullType)
```
We should update the classdoc for PrettyAttribute to reflect what it is doing.

SparkQA commented:
**[Test build #50353 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50353/consoleFull)** for PR 10757 at commit [`eef9423`](https://github.com/apache/spark/commit/eef94237ade63e03ae698870b34b620bf1c88611).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50353/
Test FAILed.

liancheng commented:
retest this please

SparkQA commented:
**[Test build #50356 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50356/consoleFull)** for PR 10757 at commit [`eef9423`](https://github.com/apache/spark/commit/eef94237ade63e03ae698870b34b620bf1c88611).

SparkQA commented:
**[Test build #50356 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50356/consoleFull)** for PR 10757 at commit [`eef9423`](https://github.com/apache/spark/commit/eef94237ade63e03ae698870b34b620bf1c88611).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50356/
Test FAILed.

liancheng commented:
retest this please

SparkQA commented:
**[Test build #50391 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50391/consoleFull)** for PR 10757 at commit [`eef9423`](https://github.com/apache/spark/commit/eef94237ade63e03ae698870b34b620bf1c88611).

SparkQA commented:
**[Test build #50391 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50391/consoleFull)** for PR 10757 at commit [`eef9423`](https://github.com/apache/spark/commit/eef94237ade63e03ae698870b34b620bf1c88611).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50391/
Test PASSed.

liancheng commented:
@rxin Thanks for the review and the summary. Will finish them in a follow-up PR.

liancheng commented:
@rxin Shall we merge this?

rxin commented:
I don't think this is blocking other work or in a code that changes frequently, so I'd prefer we just fix all the issues and have proper test coverage before merging.

liancheng committed:
Addresses PR comments

- Renames safeQuoteIdent to quoteIdentifier
- Replaces string literals and complex type extracters with
  PrettyAttributes before generating column names so that generated
  names don't contain double-quotes or back-ticks

rxin commented:
BTW You might want to resolve this before it goes super stale.

SparkQA commented:
**[Test build #51274 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51274/consoleFull)** for PR 10757 at commit [`3230d26`](https://github.com/apache/spark/commit/3230d268a49fb78f9025fbc895cfdcb36ae0e099).

SparkQA commented:
**[Test build #51274 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51274/consoleFull)** for PR 10757 at commit [`3230d26`](https://github.com/apache/spark/commit/3230d268a49fb78f9025fbc895cfdcb36ae0e099).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PrettyGetStructField(child: Expression, name: String, dataType: DataType)`
  * `case class PrettyGetArrayStructFields(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51274/
Test FAILed.

liancheng committed:
Fixes import order

SparkQA commented:
**[Test build #51286 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51286/consoleFull)** for PR 10757 at commit [`4455203`](https://github.com/apache/spark/commit/445520353007dcd0e17f7a1c1464d9c6b9507191).

SparkQA commented:
**[Test build #51286 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51286/consoleFull)** for PR 10757 at commit [`4455203`](https://github.com/apache/spark/commit/445520353007dcd0e17f7a1c1464d9c6b9507191).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51286/
Test FAILed.

SparkQA commented:
**[Test build #51296 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51296/consoleFull)** for PR 10757 at commit [`c6bf908`](https://github.com/apache/spark/commit/c6bf908ebdc3517c9db032fed3b6b1e341d4a1e1).

SparkQA commented:
**[Test build #51296 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51296/consoleFull)** for PR 10757 at commit [`c6bf908`](https://github.com/apache/spark/commit/c6bf908ebdc3517c9db032fed3b6b1e341d4a1e1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51296/
Test FAILed.

liancheng committed:
Fixes test failures

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51303/
Test FAILed.

liancheng commented:
retest this please

SparkQA commented:
**[Test build #51305 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51305/consoleFull)** for PR 10757 at commit [`0b86468`](https://github.com/apache/spark/commit/0b86468d24024b198525c86523ca63c8f5d6fa48).

SparkQA commented:
**[Test build #51305 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51305/consoleFull)** for PR 10757 at commit [`0b86468`](https://github.com/apache/spark/commit/0b86468d24024b198525c86523ca63c8f5d6fa48).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51305/
Test FAILed.

liancheng committed:
Tests for column names

SparkQA commented:
**[Test build #51307 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51307/consoleFull)** for PR 10757 at commit [`a7fe992`](https://github.com/apache/spark/commit/a7fe99247ea99b03073e353d35f4bddf443f7c8a).

SparkQA commented:
**[Test build #51307 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51307/consoleFull)** for PR 10757 at commit [`a7fe992`](https://github.com/apache/spark/commit/a7fe99247ea99b03073e353d35f4bddf443f7c8a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51307/
Test FAILed.

SparkQA commented:
**[Test build #51316 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51316/consoleFull)** for PR 10757 at commit [`11aa916`](https://github.com/apache/spark/commit/11aa916bd22d7b6c76be1192c62be7a4f173852f).

SparkQA commented:
**[Test build #2543 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2543/consoleFull)** for PR 10757 at commit [`11aa916`](https://github.com/apache/spark/commit/11aa916bd22d7b6c76be1192c62be7a4f173852f).

SparkQA commented:
**[Test build #2543 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2543/consoleFull)** for PR 10757 at commit [`11aa916`](https://github.com/apache/spark/commit/11aa916bd22d7b6c76be1192c62be7a4f173852f).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PrettyAttribute(`

SparkQA commented:
**[Test build #51316 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51316/consoleFull)** for PR 10757 at commit [`11aa916`](https://github.com/apache/spark/commit/11aa916bd22d7b6c76be1192c62be7a4f173852f).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PrettyAttribute(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51316/
Test FAILed.

liancheng committed:
Fixes PySpark test failures

SparkQA commented:
**[Test build #51337 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51337/consoleFull)** for PR 10757 at commit [`21c94d2`](https://github.com/apache/spark/commit/21c94d2224609ce3171e62c7cb58ee64cca683e7).

liancheng committed:
Fixes SparkR test cases

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51341/
Test FAILed.

liancheng commented:
retest this please

SparkQA commented:
**[Test build #51343 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51343/consoleFull)** for PR 10757 at commit [`f0eb991`](https://github.com/apache/spark/commit/f0eb9917f276a2f6f7690b9b48739d0bd2624433).

SparkQA commented:
**[Test build #51337 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51337/consoleFull)** for PR 10757 at commit [`21c94d2`](https://github.com/apache/spark/commit/21c94d2224609ce3171e62c7cb58ee64cca683e7).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51337/
Test FAILed.

SparkQA commented:
**[Test build #51343 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51343/consoleFull)** for PR 10757 at commit [`f0eb991`](https://github.com/apache/spark/commit/f0eb9917f276a2f6f7690b9b48739d0bd2624433).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51343/
Test PASSed.

SparkQA commented:
**[Test build #51349 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51349/consoleFull)** for PR 10757 at commit [`78f156c`](https://github.com/apache/spark/commit/78f156ce0ff51fbd994e15b14bda982e9fcd0868).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51348/
Test FAILed.

SparkQA commented:
**[Test build #51349 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51349/consoleFull)** for PR 10757 at commit [`78f156c`](https://github.com/apache/spark/commit/78f156ce0ff51fbd994e15b14bda982e9fcd0868).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51349/
Test FAILed.

liancheng commented:
retest this please

SparkQA commented:
**[Test build #51355 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51355/consoleFull)** for PR 10757 at commit [`78f156c`](https://github.com/apache/spark/commit/78f156ce0ff51fbd994e15b14bda982e9fcd0868).

SparkQA commented:
**[Test build #51355 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51355/consoleFull)** for PR 10757 at commit [`78f156c`](https://github.com/apache/spark/commit/78f156ce0ff51fbd994e15b14bda982e9fcd0868).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51355/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:
```diff
@@ -141,11 +144,11 @@ case class UnresolvedFunction(
-  override def prettyString: String = {
-    s"${name}(${children.map(_.prettyString).mkString(",")})"
+  override def sql: String = {
+    s"$name(${children.map(_.sql).mkString(", ")})"
```
Is this same of the default one in `Expression`? after we override `prettyName` with `name`

liancheng replied:
Yeah, I can override `prettyName` instead of `sql` here. Will update. Thanks.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:
```diff
@@ -499,6 +489,8 @@ abstract class BinaryOperator extends BinaryExpression with ExpectsInputTypes {
 
   def symbol: String
 
+  def sqlOperator: String = symbol
```
why add this concept?

liancheng replied:
For example, `And.symbol` is `&&`, while `And.sqlOperator` is `AND`. The former is used in Hive shims for converting Hive filters. The latter is used to generate SQL string.

cloud-fan replied:
makes sense, but do we override it somewhere?

liancheng replied:
It's overriden in `And` and `Or`.

liancheng replied:
Oh, although I added the method, I forgot to use it in `And` and `Or` in the revision you commented at first. Just fixed that in my last commit.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/bitwiseExpressions.scala:
```diff
@@ -123,4 +123,6 @@ case class BitwiseNot(child: Expression) extends UnaryExpression with ExpectsInp
 
   protected override def nullSafeEval(input: Any): Any = not(input)
+
+  override def sql: _root_.scala.Predef.String = s"~${child.sql}"
```
sometime IDE is stupid...

liancheng replied:
Agree...

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:
```diff
@@ -93,6 +93,8 @@ object ExtractValue {
   }
 }
 
+trait ExtractValue { this: Expression => }
```
can just be `trait ExtractValue extends Expression`

liancheng replied:
Thanks.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:
```diff
@@ -144,10 +152,11 @@ case class GetArrayStructFields(
 
   override def dataType: DataType = ArrayType(field.dataType, containsNull)
-  override def toString: String = s"$child.${field.name}"
+  override def toString: String = s"$child[$ordinal].${field.name}"
```
The javadoc of `GetArrayStructFields` is too simple...

Actually `GetArrayStructFields` means, we have an array, each element of this array is struct type, and for each element, we wanna get one field of it. It's like a batched version of `GetStructField`, just like `persons.map(_.name)` is a batched version of `person.name`.

So `s"$child.${field.name}"` is corrected here.

liancheng replied:
Ah, nice catch! Thanks!

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:
```diff
@@ -310,10 +311,20 @@ case class AttributeReference(
  */
-case class PrettyAttribute(name: String, dataType: DataType = NullType)
+case class PrettyAttribute(
+                            name: String,
```
fix indent

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:
```diff
@@ -130,19 +133,28 @@ package object util {
+  def usePrettyExpression(e: Expression): Expression = e transform {
+    case a: Attribute => new PrettyAttribute(a)
+    case Literal(s: UTF8String, StringType) => PrettyAttribute(s.toString, StringType)
+    case e @ GetStructField(child, _, Some(name)) => PrettyGetStructField(child, name, e.dataType)
```
can we also turn `GetStructField` to `PrettyAttribute` here? We can generate the sql string here and pass it as the name of `PrettyAttribute`

liancheng replied:
Probably not. Because identifiers in `GetStructField.sql` are quoted.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonUDF.scala:
```diff
@@ -36,9 +36,12 @@ case class PythonUDF(
 
   override def nullable: Boolean = true
+
+  override def sql: String = s"$name(${children.mkString(", ")})"
```
It extends `NonSQLExpression`, no need to implement it?

liancheng replied:
`Expression.sql` is also called when generating column names. `NonSQLExpressoin.sql` calls `PythonUDF.toString` here, which returns `PythonUDF#...`. We probably don't want to have `#` appear in column names.

cloud-fan replied:
So it doesn't need to extend `NonSQLExpression`?

liancheng replied:
Well, it does. Because we can't generate SQL string for a query plan containing a Python UDF. `SQLBuilder` tries to find `NonSQLExpression`s for this detection.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:
```diff
@@ -255,6 +236,19 @@ trait Unevaluable extends Expression {
+ * `ScalaUDF`, `ScalaUDAF`, and object expressions like `MapObjects` and `Invoke`.
+ */
+trait NonSQLExpression extends Expression {
+  override def sql: String = {
```
how about mark it as final?

liancheng replied:
One exception is the `PythonUDF` case mentioned above.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala:
```diff
@@ -37,11 +38,21 @@ import org.apache.spark.sql.execution.datasources.LogicalRelation
  * supported by this builder (yet).
  */
 class SQLBuilder(logicalPlan: LogicalPlan, sqlContext: SQLContext) extends Logging {
+  require(logicalPlan.resolved, "SQLBuilder only supports reslved logical query plans")
```
nit: `reslved` -> `resolved`

liancheng replied:
Thanks.

liancheng committed:
Addresses PR comments

SparkQA commented:
**[Test build #51364 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51364/consoleFull)** for PR 10757 at commit [`c40862f`](https://github.com/apache/spark/commit/c40862fdf8a634c8f4f91353cdec309d3e4d0fef).

SparkQA commented:
**[Test build #51364 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51364/consoleFull)** for PR 10757 at commit [`c40862f`](https://github.com/apache/spark/commit/c40862fdf8a634c8f4f91353cdec309d3e4d0fef).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51364/
Test PASSed.

liancheng commented:
@rxin I think this one is ready for sign off now.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:
```diff
@@ -130,12 +134,17 @@ case class GetStructField(child: Expression, ordinal: Int, name: Option[String]
+  extends UnaryExpression with Unevaluable {
 
-  override def sql: String = child.sql + s".`${childSchema(ordinal).name}`"
+  override def sql: String = s"${child.sql}.$name"
```
can we use this string as the name of `PrettyAttribute`? Then we can avoid creating these classes.

liancheng replied:
No, `child.sql` may contain back-ticks/double-quotes.

cloud-fan replied:
what if we use `transformUp` in `usePrettyExpression`, so that we can make sure when we generate this string, `child.sql` won't have back-ticks/double-quotes?

liancheng replied:
Good idea. Updated and removed these new classes.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:
```diff
@@ -310,10 +311,20 @@ case class AttributeReference(
+case class PrettyAttribute(
+    name: String,
+    dataType: DataType = NullType,
+    override val foldable: Boolean = false)
```
Is there some case we need the foldable to be true?

liancheng replied:
Oh, this is an unexpected change. I should revert it.

cloud-fan commented:
My last concern is changing default column alias `_c0, _c1, ..` to `Expression.sql`. Before this PR, query plans generated by SQL string will use `_c0, _c1, ..`, query plans generated by DataFrame will use `prettyString`. Now we always use `Expression.sql`, this is more consistent, but will break some old code, e.g. `select _c1 from (select a + 1, b + 2 from tbl) t `, does it worth?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:
```diff
@@ -198,18 +208,26 @@ case class GetArrayStructFields(
+    child: Expression, ordinal: Int, name: String, dataType: DataType)
+  extends UnaryExpression with Unevaluable {
+
+  override def sql: String = s"${child.sql}[$ordinal].$name"
```
should be `s"${child.sql}.$name"`?

liancheng replied:
Yes, thanks!

liancheng commented:
@cloud-fan I'm open to the new column naming issue. Personally I tend to make it more consistent. Please note that replace `prettyString` with `sql` already breaks compatibility (e.g. `a && b` is renamed to `a AND b`).

SparkQA commented:
**[Test build #51411 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51411/consoleFull)** for PR 10757 at commit [`bce3339`](https://github.com/apache/spark/commit/bce3339f42bebf98092a4e67d3d8028d27624704).

liancheng committed:
Test case for GetArrayStructFields

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51412/
Test FAILed.

liancheng commented:
retest this please

SparkQA commented:
**[Test build #51417 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51417/consoleFull)** for PR 10757 at commit [`e9f9fd9`](https://github.com/apache/spark/commit/e9f9fd9208e4729dcac58ec66c5d542e322af9eb).

SparkQA commented:
**[Test build #51411 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51411/consoleFull)** for PR 10757 at commit [`bce3339`](https://github.com/apache/spark/commit/bce3339f42bebf98092a4e67d3d8028d27624704).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51411/
Test FAILed.

SparkQA commented:
**[Test build #51417 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51417/consoleFull)** for PR 10757 at commit [`e9f9fd9`](https://github.com/apache/spark/commit/e9f9fd9208e4729dcac58ec66c5d542e322af9eb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51417/
Test PASSed.

rxin commented:
I likely won't have time to look at this until next week. Feel free to merge if @cloud-fan lgtms. I can do some post-hoc review later.

liancheng committed:
Code simplification

- Removes PrettyGetStructField and PrettyGetArrayStructFields
- Using literal suffixes in literal SQL representations

SparkQA commented:
**[Test build #51480 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51480/consoleFull)** for PR 10757 at commit [`587a69c`](https://github.com/apache/spark/commit/587a69cf957484793abb5d70a41f4ce097db796e).

SparkQA commented:
**[Test build #51480 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51480/consoleFull)** for PR 10757 at commit [`587a69c`](https://github.com/apache/spark/commit/587a69cf957484793abb5d70a41f4ce097db796e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51480/
Test FAILed.

SparkQA commented:
**[Test build #51485 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51485/consoleFull)** for PR 10757 at commit [`8a2a3a8`](https://github.com/apache/spark/commit/8a2a3a8c886b1aa44b132a4e92fdc59496d78c1c).

SparkQA commented:
**[Test build #51485 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51485/consoleFull)** for PR 10757 at commit [`8a2a3a8`](https://github.com/apache/spark/commit/8a2a3a8c886b1aa44b132a4e92fdc59496d78c1c).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51485/
Test FAILed.

SparkQA commented:
**[Test build #51493 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51493/consoleFull)** for PR 10757 at commit [`e5eec09`](https://github.com/apache/spark/commit/e5eec09815e6f9053ad492981711388d7ec62bc0).

SparkQA commented:
**[Test build #51493 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51493/consoleFull)** for PR 10757 at commit [`e5eec09`](https://github.com/apache/spark/commit/e5eec09815e6f9053ad492981711388d7ec62bc0).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51493/
Test FAILed.

cloud-fan commented:
LGTM as long as the tests pass.

liancheng commented:
Thanks for the detailed review!

liancheng committed:
No numeric suffixes for numeric literals in column names

liancheng commented:
One more update after offline discussion with @rxin: numeric literals appearing in column names shouldn't come with numeric suffixes or casts. This is done by replacing numeric literals with `PrettyAttribute` in `usePrettyExpression`.

SparkQA commented:
**[Test build #51555 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51555/consoleFull)** for PR 10757 at commit [`a6adbdb`](https://github.com/apache/spark/commit/a6adbdb91fce157eba0cbe22cd42164d7f505d45).

SparkQA commented:
**[Test build #51555 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51555/consoleFull)** for PR 10757 at commit [`a6adbdb`](https://github.com/apache/spark/commit/a6adbdb91fce157eba0cbe22cd42164d7f505d45).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51555/
Test FAILed.

SparkQA commented:
**[Test build #51591 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51591/consoleFull)** for PR 10757 at commit [`6723ab6`](https://github.com/apache/spark/commit/6723ab64f95b9f854f6a9a3c33ec81b5f6c454ae).

SparkQA commented:
**[Test build #51591 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51591/consoleFull)** for PR 10757 at commit [`6723ab6`](https://github.com/apache/spark/commit/6723ab64f95b9f854f6a9a3c33ec81b5f6c454ae).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51591/
Test FAILed.

SparkQA commented:
**[Test build #51619 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51619/consoleFull)** for PR 10757 at commit [`e21fded`](https://github.com/apache/spark/commit/e21fded9cff0c2ff47b59d1c71b6790a8b4d3f46).

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51615/
Test FAILed.

SparkQA commented:
**[Test build #51619 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51619/consoleFull)** for PR 10757 at commit [`e21fded`](https://github.com/apache/spark/commit/e21fded9cff0c2ff47b59d1c71b6790a8b4d3f46).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51619/
Test FAILed.

liancheng committed:
Fixes test failure

SparkQA commented:
**[Test build #51626 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51626/consoleFull)** for PR 10757 at commit [`bb65e4d`](https://github.com/apache/spark/commit/bb65e4d7b31a6e4570e19d6179ffb6f9c8b813b2).

SparkQA commented:
**[Test build #51626 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51626/consoleFull)** for PR 10757 at commit [`bb65e4d`](https://github.com/apache/spark/commit/bb65e4d7b31a6e4570e19d6179ffb6f9c8b813b2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51626/
Test PASSed.

liancheng commented:
OK I'm going to merge this to master.
